added flag for retrieving parameters from the command line

hjkatz opened the pull request:
- Updated the arguments to be passed as a hash for simplicity (all the defaults have been kept the same)
- Added new flag `--parameters-cli` which will prompt the user for parameters values at runtime

jonaf commented on lib/cloudformation-ruby-dsl/cfntemplate.rb:
```diff
@@ -293,6 +299,9 @@ def cfn(template)
       create_stack_opts = extra_options.merge(create_stack_opts)
 
+      # remove custom options
+      create_stack_opts.delete(:parameters_cli)
```
Would this need to also be done for validate, update, etc.? Is the `--parameters-cli` option available for any of these subcommands?

hjkatz replied:
How the application is structured has the actual evaluation of the parameters at runtime during the dsl parsing. So the actual cli portion is run when you call `parameter ...` in the template block. This is a bit weird of a place, but it should allow for validate, update, etc... anything that parses the template.

A feature that could possibly be added is actual validation using the AWS SDK for each param as they are entered. But, I think it's good enough to have a easy way to enter the params for now.

jonaf commented on lib/cloudformation-ruby-dsl/dsl.rb:
```diff
@@ -71,7 +77,78 @@ def exec!()
-    @parameters[name] ||= options[:Default]
+
+    if @parameters_cli
+      @parameters[name] ||= _get_parameter_from_cli(name, options)
```
Maybe it's too hairy to get into this, but it would be kind of neat if I were doing an update to a stack and this either printed the current value (so I can just re-use it) or offered me the option "Use current value?" if a current value was already set for that parameter in the stack (excluding the default).

hjkatz replied:
mmmm, this might be possible. I'll take a look and see.

jonaf commented on lib/cloudformation-ruby-dsl/cfntemplate.rb:
```diff
@@ -70,26 +70,32 @@ def to_json(*a)
-      parameters = Hash[value.split(/;/).map { |pair| pair.split(/=/, 2) }]  #/# fix for syntax highlighting
+      args[:parameters] = Hash[value.split(/;/).map { |pair| pair.split(/=/, 2) }]  #/# fix for syntax highlighting
+    when '--parameters-cli'
+      args[:parameters_cli] = true
```
What if we called this `--interactive` / `-i`? Maybe that's too generic. `--interactive-parameters` or similar? What do you think?

hjkatz replied:
The name was a bit tricky to pick, whatever you want is fine with me.

hjkatz commented:
Please review the recent changes.

- changed flag name from `--parameters-cli` to `--interactive` (hopefully for a more encompassed interactive mode in the future)
- fixed a bug with custom parameters in the update step (similar to the create, but I forgot initially)
- decided not to do the additional "Use previous value" for parameters during the update

I looked into the amount of effort for attempting to use the old stack during an update step to get the differences in parameters interactively, but it proved too high (for now). It would require a restructuring with a more direct call to dsl compile or something similar. This is because the params are parsed immediately as they are called in the dsl, instead of parsing the dsl then asking for params. This ordering makes it very difficult to even add the interactive mode in the first place.

I would recommend a restructuring of the `cfntemplate.rb` file into 2 files. One for running, one for the template. Then you can add a compile/parse/extract step for all cli things first (such as params, tags, etc...), and after that you can use those parsed values to create the final template.
